backends/p4tools/testgen/targets/bmv2/expr_stepper.cpp

```diff
@@ -0,0 +1,1742 @@
+             nextState->add(new TraceEvent::Generic("mark_to_drop executed."));
+             nextState->popBody();
+             result->emplace_back(nextState);
+         }},
+        /* ======================================================================================
+         *  random
+         *  Generate a random number in the range lo..hi, inclusive, and write
+         *  it to the result parameter.
+         * ====================================================================================== */
+        {"*method.random",
+         {"result", "lo", "hi"},
+         [this](const IR::MethodCallExpression* /*call*/, const IR::Expression* /*receiver*/,
+                IR::ID& /*methodName*/, const IR::Vector<IR::Argument>* args,
+                const ExecutionState& state, SmallStepEvaluator::Result& result) {
+             BUG_CHECK(args->at(1)->expression->to<IR::Constant>(), "Expected a constant.");
+             BUG_CHECK(args->at(2)->expression->to<IR::Constant>(), "Expected a constant.");
+             const auto* lo = args->at(1)->expression->to<IR::Constant>();
+             const auto* hi = args->at(2)->expression->to<IR::Constant>();
+             BUG_CHECK(lo->value <= hi->value,
+                       "Low value ( %1% ) must be less than high value ( %2% ).", lo, hi);
+             auto* nextState = new ExecutionState(state);
+             const auto* resultField = args->at(0)->expression;
+             const IR::Member* fieldRef = nullptr;
+             if (const auto* pathRef = resultField->to<IR::PathExpression>()) {
+                 fieldRef = state.convertPathExpr(pathRef);
+             } else {
+                 fieldRef = resultField->to<IR::Member>();
+             }
+             if (fieldRef == nullptr) {
+                 TESTGEN_UNIMPLEMENTED("Random output %1% of type %2% not supported", resultField,
+                                       resultField->type);
+             }
+
+             // If the range is limited to only one value, return that value.
+             if (lo->value == hi->value) {
+                 nextState->set(fieldRef, hi);
+                 nextState->popBody();
+                 result->emplace_back(nextState);
+                 return false;
+             }
+             // Otherwise, we will have to return taint,
+             // as we do not control the random generator of the device under test.
+             if (resultField->type->is<IR::Type_Bits>()) {
+                 nextState->set(fieldRef,
+                                programInfo.createTargetUninitialized(resultField->type, true));
+                 nextState->popBody();
+                 result->emplace_back(nextState);
+                 return false;
+             }
+             BUG("Not a Type_Bits: %1%", resultField->type);
+         }},
+        /* ======================================================================================
+         *  assume
```

Comment:
this is great documentation, it should not be only in code, it should be in some documents that users can read.

Reply:
This documentation is actually part of the v1model.p4 file, and most of this documentation was written by @jafingerhut. I do not want to take credit for that. :) 

The documentation for v1model was actually very helpful for implementing the semantics of the architecture.